### PR TITLE
JIT: Make optAssertionPropGlobal_RelOp less conservative for side-effects

### LINKED ISSUE #85522: JIT: suboptimal codegen for isinst + cast pattern
```csharp
int Test(object o)
{
    if (o is string)
    {
        return ((string)o).Length;
    }
    return 0;
}
```
Current codegen:
```asm
; Method Proga:Test(System.Object):int
       mov      rax, rcx
       test     rax, rax
       je       SHORT G_M10539_IG06
       mov      rdx, 0x7FF862F27778      ; System.String
       cmp      qword ptr [rax], rdx
       jne      SHORT G_M10539_IG06
       mov      rax, 0x7FF862F27778      ; System.String
       cmp      qword ptr [rcx], rax
       jne      SHORT G_M10539_IG08
       mov      eax, dword ptr [rcx+08H]
       ret      
G_M10539_IG06:
       xor      eax, eax
       ret      
G_M10539_IG08:
       int3     ;;; where is the throw block ???
; Total bytes of code: 46
```
Type is checked twice.

It doesn't happen with the pattern-match syntax:
```csharp
int Test(object o)
{
    if (o is string s)
    {
        return s.Length;
    }
    return 0;
}
```

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/85522

```csharp
static int Test(object o)
{
    if (o is string)
        return ((string)o).Length;
    return 0;
}
```
Old codegen:
```asm
; Method Proga:Test(System.Object):int
       mov      rax, rcx
       test     rax, rax
       je       SHORT G_M10539_IG06
       mov      rdx, 0x7FF85AF27778      ; System.String
       cmp      qword ptr [rax], rdx
       jne      SHORT G_M10539_IG06
       mov      rax, 0x7FF85AF27778      ; System.String
       cmp      qword ptr [rcx], rax
       jne      SHORT G_M10539_IG08
       mov      eax, dword ptr [rcx+08H]
       ret      
G_M10539_IG06:  
       xor      eax, eax
       ret      
G_M10539_IG08:  
       int3     
; Total bytes of code: 46
```
New codegen:
```asm
; Method Proga:Test(System.Object):int
       mov      rax, rcx
       test     rax, rax
       je       SHORT G_M10539_IG04
       mov      rdx, 0x7FF85AF37778      ; System.String
       cmp      qword ptr [rax], rdx
       je       SHORT G_M10539_IG06
G_M10539_IG04:
       xor      eax, eax
       ret      
G_M10539_IG06:  
       mov      eax, dword ptr [rcx+08H]
       ret      
; Total bytes of code: 30
```

The problem that for the following tree

image

Assertprop gives up due to ASG side-effects from the op2 (op1 is side-effect free since it's proved that the address is not null) - it will be preserved just fine

### CODE REVIEW

**Comment by EgorBo** [comment]
Some free perf, [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=256191&view=ms.vss-build-web.run-extensions-tab)

@dotnet/jit-contrib @AndyAyersMS @jakobbotsch @SingleAccretion PTAL, `optAssertionPropGlobal_RelOp` only bashes op1 to a constant (op2 only if it's a local) so a possible side effect is kept alive after foldings

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -3802,8 +3802,8 @@ GenTree* Compiler::optAssertionPropGlobal_RelOp(ASSERT_VALARG_TP assertions, Gen
         return nullptr;
     }
 
-    // Bail out if tree is not side effect free.
-    if ((tree->gtFlags & GTF_SIDE_EFFECT) != 0)
+    // Bail out if op1 is not side effect free.
+    if ((tree->gtGetOp1()->gtFlags & GTF_SIDE_EFFECT) != 0)
```
Comment (SingleAccretion) [inline src/coreclr/jit/assertionprop.cpp:3806]:
```suggestion
    // Bail out if op1 is not side effect free. Note we'll be bashing it below, unlike op2.
    if ((op1->gtFlags & GTF_SIDE_EFFECT) != 0)
```

**Review by SingleAccretion — COMMENTED** [review]
Side note: the way this method works seems more complicated than it needs be.

**Comment by EgorBo** [comment]
> Side note: the way this method works seems more complicated than it needs be.

Agree, left a note for myself to refactor it, but in this PR just wanted to fix an annoying CQ issue 🙂

**Comment by EgorBo** [comment]
@jakobbotsch [the only](https://dev.azure.com/dnceng-public/public/_build/results?buildId=252520&view=ms.vss-build-web.run-extensions-tab) fuzzlyn failure doesn't look to be assertprop, does it look familiar?

**Comment by jakobbotsch** [comment]
> @jakobbotsch [the only](https://dev.azure.com/dnceng-public/public/_build/results?buildId=256662&view=ms.vss-build-web.run-extensions-tab) fuzzlyn failure doesn't look to be assertprop, does it look familiar?

Hmm no, I don't recognize that one.

**Review by jakobbotsch — APPROVED** [review]

**Comment by EgorBo** [comment]
I'll check the failure on my osx-arm64 if it repros and file an issue

**Pull request merged**